src/unix/uiactionx11.cpp

```diff
@@ -46,6 +55,113 @@ GtkWidget* wxGetTopLevelGTK();
 
 namespace
 {
+// Like the real events, this class tries to add a _fake_ delay to the generated
+// (fake) events so that chances that they got lost or ignored (for whatever reasons)
+// by the X server (or the WM, or even the input device driver) are minimized.
+class wxXSync
+{
+public:
+    wxXSync(wxX11Display& display)
+        : m_display(display), m_isMotion(true)
+    {
+    }
+
+    wxXSync(wxX11Display& display, bool depressed)
+        : m_display(display), m_isMotion(false)
+    {
+        static int s_maxDepressed = 0;
+
+        if ( depressed )
+        {
+            ++ms_numDepressed; // A key or button is down.
+
+            if ( ms_numDepressed - s_maxDepressed != 1 )
+                wxFAIL_MSG("Invalid call to wxXSync() ctor");
+
+            s_maxDepressed = ms_numDepressed;
+        }
+        else
+        {
+            if ( s_maxDepressed != 0 )
+            {
+                s_maxDepressed = 0; // reset
+
+                wxYield();
+                wxMilliSleep(Default_Delay);
+            }
+
+            --ms_numDepressed; // A key or button is up.
+
+            if ( ms_numDepressed < 0 )
+                wxFAIL_MSG("Invalid call to wxXSync() ctor");
+        }
+    }
+
+    ~wxXSync()
+    {
+        XSync(m_display, False);
+
+        m_isMotion ? OnMotion() : OnKeyOrButtonUp();
+    }
+
+private:
+    static void OnMotion()
+    {
+        wxMilliSleep(Motion_Delay);
+        wxYield();
+    }
+
+    static void OnKeyOrButtonUp()
+    {
+        if ( ms_numDepressed > 0 )
+        {
+            // Do nothing if a key / button is still depressed.
+            return;
+        }
+
+        wxYield();
+
+        // Default_Delay might not be sufficient for some
+        // tests which need more additionnal time to pass.
+        wxMilliSleep(Default_Delay + GetAdditionalDelay());
+    }
+
+    static long GetAdditionalDelay()
```

Comment:
Passing extra delay from one part of the program to another via environment variables is rather strange. Why not just have some public `static void SetAdditionalDelay(long delay)` instead and call it from the testing code?

Also, why should we set _additional_ delay instead of just setting the delay? I feel like this is more complicated than necessary and we should just have some `ms_delay`, initialized with `Default_Delay`, and then functions for getting/setting it.

Reply:
I agree with you that this is a poor design choice and should be change to something better. other than this, **KeyboardEventTestCase** (tests/events/keyboard.cpp) is the only place where the `Default_Delay` is not sufficient and need more additional delay to pass successfully, and IMHO it doesn't worth the trouble to add any new API (at least for now) which won't be needed for the majority of test cases. what do you think ?

Reply:
The problem is that this env var is also a kind of API, just a hidden and clumsy one, and it doesn't seem preferable to me to just adding the functions in any way.

But looking at this more closely, could we perhaps just always add 10ms for the key events if this is what they need? AFAICS this would make things simpler and what exactly is the drawback of doing it? Just slightly longer test run-time? I think we can live with it...

Reply:
I'll try to get rid of it then...